README
	slp - Simplify the writing of beamer tex slides by providing shorthand macros



  Copyright (C) 2012 Reiner Jung

  slp is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Comments are welcome.

	- Reiner Jung <rju@informatik.uni-kiel.de>

Simple language description:

The processor work with different interpretation modes. On start it is in
top mode. Top mode allows the definition of title, date, author, part, sections
and frames. Whitespaces and newlines are ignored.

To define a title write

title: TITLE-TEXT

TITLE-TEXT is a char sequence which ends with a newline \n

The commands date:, author: work similar. They can be used to define the author
and the presentation date.

The command part: followed by a char sequence (ends with newline) creates an
\part{VALUE} in the output.

The command sec: followed by a char sequence (ends with newline) creates an
\section{VALUE} in the output.

A frame is started with >. For example

> MY-FRAME-TITLE

The must be a whitespace after the > and then the frame title is the following
text until the newline.

Inside a frame another set of commands is valid. In that context structures can
be defined with a line starting with '

'My Structure

which generates to \structure{VALUE}

Images are created with img:[properties]file-reference





ChangeLog
2012-04-18  Reiner Jung  <rju@informatik.uni-kiel.de>

	* slp: initial version.



slp.c
/* 
 slp - Simplify the writing of beamer tex slides by providing shorthand macros

 Copyright (C) 2012 Reiner Jung

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2, or (at your option)
 any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software Foundation,
 Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

 #include <termios.h>
 #include <grp.h>
 #include <pwd.h>
 */

#include <stdio.h>
#include <sys/types.h>
#include <argp.h>
#include "system.h"
#include "scanner.h"
#include "parser.h"

#define EXIT_FAILURE 1

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# define textdomain(Domain)
# define _(Text) Text
#endif
#define N_(Text) Text

char *xmalloc();
char *xrealloc();
char *xstrdup();

static error_t parse_opt(int key, char *arg, struct argp_state *state);
static void show_version(FILE *stream, struct argp_state *state);

/* argp option keys */
enum {
	DUMMY_KEY = 129
};

/* Option flags and variables.  These are initialized in parse_opt.  */

char *oname; /* --output=FILE */

static struct argp_option options[] = { { "output", 'o', N_("FILE"), 0,
		N_("Send output to FILE instead of standard output"), 0 }, { NULL,
		0, NULL, 0, NULL, 0 } };

/* The argp functions examine these global variables.  */
const char *argp_program_bug_address = "<rju@informatik.uni-kiel.de>";
void (*argp_program_version_hook)(FILE *, struct argp_state *) = show_version;

static struct argp argp =
		{ options, parse_opt, N_("[FILE...]"),
				N_("Simplify the writing of beamer tex slides by providing shorthand macros"),
				NULL, NULL, NULL };

int main(int argc, char **argv) {
	yyin = NULL;
	textdomain(PACKAGE);
	argp_parse(&argp, argc, argv, 0, NULL, NULL);

	if (yyin != NULL) {
		parse();
		fclose(yyin);
	}

	exit(0);
}

/* Parse a single option.  */
static error_t parse_opt(int key, char *arg, struct argp_state *state) {
	switch (key) {
	case ARGP_KEY_INIT:
		/* Set up default values.  */
		oname = "stdout";
		ofile = stdout;
		break;

	case 'o': /* --output */
		oname = xstrdup(arg);
		ofile = fopen(oname, "w");
		if (!ofile)
			argp_failure(state, EXIT_FAILURE, errno,
					_("Cannot open %s for writing"), oname);
		break;

	case ARGP_KEY_ARG: /* [FILE]... */
		if ((yyin = fopen(arg, "r")) == 0) {
			fprintf(stderr, "File not found: %s\n", arg);
			exit(1);
		}
		break;

	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

/* Show the version number and copyright information.  */
static void show_version(FILE *stream, struct argp_state *state) {
	(void) state;
	/* Print in small parts whose localizations can hopefully be copied
	 from other programs.  */
	fputs(PACKAGE" "VERSION"\n", stream);
	fprintf(stream, _("Written by %s.\n\n"), "Reiner Jung");
	fprintf(stream, _("Copyright (C) %s %s\n"), "2012", "Reiner Jung");
	fputs(
			_("\
This program is free software; you may redistribute it under the terms of\n\
the GNU General Public License.  This program has absolutely no warranty.\n"),
			stream);
}


parser.h
/*
 * parser.h
 *
 *  Created on: 18 Apr 2012
 *      Author: rju
 */

#ifndef PARSER_H_
#define PARSER_H_

#include <stdio.h>
#include <stdlib.h> 

#define TITLE        101
#define AUTHOR       102
#define DATE         103
#define SECTION      104
#define IMAGE        105
#define LISTING      106
#define ID           107
#define SEP          108
#define LABEL        109
#define LINE         110
#define VALUE        111
#define LSTEND       112
#define PSEQ_START   113
#define PSEQ_END     114
#define FRAME        115
#define STRUCT       116
#define ITEM         117
#define NEWLINE      118
#define INDENT       119
#define SPACE        120
#define UNCOVER      121
#define OVERLAY      122
#define ONLY         123
#define URL          124
#define COLUMN       125
#define COLUMN_SEP   126
#define END          127
#define OVERLAY_CODE 128
#define PROPERTY     129
#define PROP_SEP     130
#define PART         131
#define HEAD_END     132
#define DESC_SEP     133

#define ZERO         0

extern char* data;

extern char *item_type;
extern FILE *ofile;

int parse();
char* strtrim(const char *string);

#endif /* PARSER_H_ */


scanner.h
#ifndef yyHEADER_H
#define yyHEADER_H 1
#define yyIN_HEADER 1

#line 6 "scanner.h"

#define  YY_INT_ALIGNED short int

/* A lexical scanner generated by flex */

#define FLEX_SCANNER
#define YY_FLEX_MAJOR_VERSION 2
#define YY_FLEX_MINOR_VERSION 5
#define YY_FLEX_SUBMINOR_VERSION 35
#if YY_FLEX_SUBMINOR_VERSION > 0
#define FLEX_BETA
#endif

/* First, we deal with  platform-specific or compiler-specific issues. */

/* begin standard C headers. */
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>

/* end standard C headers. */

/* flex integer type definitions */

#ifndef FLEXINT_H
#define FLEXINT_H

/* C99 systems have <inttypes.h>. Non-C99 systems may or may not. */

#if defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L

/* C99 says to define __STDC_LIMIT_MACROS before including stdint.h,
 * if you want the limit (max/min) macros for int types. 
 */
#ifndef __STDC_LIMIT_MACROS
#define __STDC_LIMIT_MACROS 1
#endif

#include <inttypes.h>
typedef int8_t flex_int8_t;
typedef uint8_t flex_uint8_t;
typedef int16_t flex_int16_t;
typedef uint16_t flex_uint16_t;
typedef int32_t flex_int32_t;
typedef uint32_t flex_uint32_t;
#else
typedef signed char flex_int8_t;
typedef short int flex_int16_t;
typedef int flex_int32_t;
typedef unsigned char flex_uint8_t; 
typedef unsigned short int flex_uint16_t;
typedef unsigned int flex_uint32_t;

/* Limits of integral types. */
#ifndef INT8_MIN
#define INT8_MIN               (-128)
#endif
#ifndef INT16_MIN
#define INT16_MIN              (-32767-1)
#endif
#ifndef INT32_MIN
#define INT32_MIN              (-2147483647-1)
#endif
#ifndef INT8_MAX
#define INT8_MAX               (127)
#endif
#ifndef INT16_MAX
#define INT16_MAX              (32767)
#endif
#ifndef INT32_MAX
#define INT32_MAX              (2147483647)
#endif
#ifndef UINT8_MAX
#define UINT8_MAX              (255U)
#endif
#ifndef UINT16_MAX
#define UINT16_MAX             (65535U)
#endif
#ifndef UINT32_MAX
#define UINT32_MAX             (4294967295U)
#endif

#endif /* ! C99 */

#endif /* ! FLEXINT_H */

#ifdef __cplusplus

/* The "const" storage-class-modifier is valid. */
#define YY_USE_CONST

#else	/* ! __cplusplus */

/* C99 requires __STDC__ to be defined as 1. */
#if defined (__STDC__)

#define YY_USE_CONST

#endif	/* defined (__STDC__) */
#endif	/* ! __cplusplus */

#ifdef YY_USE_CONST
#define yyconst const
#else
#define yyconst
#endif

/* Size of default input buffer. */
#ifndef YY_BUF_SIZE
#ifdef __ia64__
/* On IA-64, the buffer size is 16k, not 8k.
 * Moreover, YY_BUF_SIZE is 2*YY_READ_BUF_SIZE in the general case.
 * Ditto for the __ia64__ case accordingly.
 */
#define YY_BUF_SIZE 32768
#else
#define YY_BUF_SIZE 16384
#endif /* __ia64__ */
#endif

#ifndef YY_TYPEDEF_YY_BUFFER_STATE
#define YY_TYPEDEF_YY_BUFFER_STATE
typedef struct yy_buffer_state *YY_BUFFER_STATE;
#endif

extern int yyleng;

extern FILE *yyin, *yyout;

#ifndef YY_TYPEDEF_YY_SIZE_T
#define YY_TYPEDEF_YY_SIZE_T
typedef size_t yy_size_t;
#endif

#ifndef YY_STRUCT_YY_BUFFER_STATE
#define YY_STRUCT_YY_BUFFER_STATE
struct yy_buffer_state
	{
	FILE *yy_input_file;

	char *yy_ch_buf;		/* input buffer */
	char *yy_buf_pos;		/* current position in input buffer */

	/* Size of input buffer in bytes, not including room for EOB
	 * characters.
	 */
	yy_size_t yy_buf_size;

	/* Number of characters read into yy_ch_buf, not including EOB
	 * characters.
	 */
	int yy_n_chars;

	/* Whether we "own" the buffer - i.e., we know we created it,
	 * and can realloc() it to grow it, and should free() it to
	 * delete it.
	 */
	int yy_is_our_buffer;

	/* Whether this is an "interactive" input source; if so, and
	 * if we're using stdio for input, then we want to use getc()
	 * instead of fread(), to make sure we stop fetching input after
	 * each newline.
	 */
	int yy_is_interactive;

	/* Whether we're considered to be at the beginning of a line.
	 * If so, '^' rules will be active on the next match, otherwise
	 * not.
	 */
	int yy_at_bol;

    int yy_bs_lineno; /**< The line count. */
    int yy_bs_column; /**< The column count. */
    
	/* Whether to try to fill the input buffer when we reach the
	 * end of it.
	 */
	int yy_fill_buffer;

	int yy_buffer_status;

	};
#endif /* !YY_STRUCT_YY_BUFFER_STATE */

void yyrestart (FILE *input_file  );
void yy_switch_to_buffer (YY_BUFFER_STATE new_buffer  );
YY_BUFFER_STATE yy_create_buffer (FILE *file,int size  );
void yy_delete_buffer (YY_BUFFER_STATE b  );
void yy_flush_buffer (YY_BUFFER_STATE b  );
void yypush_buffer_state (YY_BUFFER_STATE new_buffer  );
void yypop_buffer_state (void );

YY_BUFFER_STATE yy_scan_buffer (char *base,yy_size_t size  );
YY_BUFFER_STATE yy_scan_string (yyconst char *yy_str  );
YY_BUFFER_STATE yy_scan_bytes (yyconst char *bytes,int len  );

void *yyalloc (yy_size_t  );
void *yyrealloc (void *,yy_size_t  );
void yyfree (void *  );

/* Begin user sect3 */

extern int yylineno;

extern char *yytext;
#define yytext_ptr yytext

#ifdef YY_HEADER_EXPORT_START_CONDITIONS
#define INITIAL 0
#define value 1
#define lsthead 2
#define lst 3

#endif

#ifndef YY_NO_UNISTD_H
/* Special case for "unistd.h", since it is non-ANSI. We include it way
 * down here because we want the user's section 1 to have been scanned first.
 * The user has a chance to override it with an option.
 */
#include <unistd.h>
#endif

#ifndef YY_EXTRA_TYPE
#define YY_EXTRA_TYPE void *
#endif

/* Accessor methods to globals.
   These are made visible to non-reentrant scanners for convenience. */

int yylex_destroy (void );

int yyget_debug (void );

void yyset_debug (int debug_flag  );

YY_EXTRA_TYPE yyget_extra (void );

void yyset_extra (YY_EXTRA_TYPE user_defined  );

FILE *yyget_in (void );

void yyset_in  (FILE * in_str  );

FILE *yyget_out (void );

void yyset_out  (FILE * out_str  );

int yyget_leng (void );

char *yyget_text (void );

int yyget_lineno (void );

void yyset_lineno (int line_number  );

/* Macros after this point can all be overridden by user definitions in
 * section 1.
 */

#ifndef YY_SKIP_YYWRAP
#ifdef __cplusplus
extern "C" int yywrap (void );
#else
extern int yywrap (void );
#endif
#endif

#ifndef yytext_ptr
static void yy_flex_strncpy (char *,yyconst char *,int );
#endif

#ifdef YY_NEED_STRLEN
static int yy_flex_strlen (yyconst char * );
#endif

#ifndef YY_NO_INPUT

#endif

/* Amount of stuff to slurp up with each read. */
#ifndef YY_READ_BUF_SIZE
#ifdef __ia64__
/* On IA-64, the buffer size is 16k, not 8k */
#define YY_READ_BUF_SIZE 16384
#else
#define YY_READ_BUF_SIZE 8192
#endif /* __ia64__ */
#endif

/* Number of entries by which start-condition stack grows. */
#ifndef YY_START_STACK_INCR
#define YY_START_STACK_INCR 25
#endif

/* Default declaration of generated scanner - a define so the user can
 * easily add parameters.
 */
#ifndef YY_DECL
#define YY_DECL_IS_OURS 1

extern int yylex (void);

#define YY_DECL int yylex (void)
#endif /* !YY_DECL */

/* yy_get_previous_state - get the state just before the EOB char was reached */

#undef YY_NEW_FILE
#undef YY_FLUSH_BUFFER
#undef yy_set_bol
#undef yy_new_buffer
#undef yy_set_interactive
#undef YY_DO_BEFORE_ACTION

#ifdef YY_DECL_IS_OURS
#undef YY_DECL_IS_OURS
#undef YY_DECL
#endif

#line 82 "scanner.l"


#line 332 "scanner.h"
#undef yyIN_HEADER
#endif /* yyHEADER_H */


parser.c
/*
 * parser.c
 *
 *  Created on: 18 Apr 2012
 *      Author: rju
 */

#include "parser.h"
#include "scanner.h"
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>

/* list of token names and descriptions */
char *token_names[] = {
	"title command","author command","date command","section command",
	"image command","embedded listing",
	"ID","separator","label","LINE","value",
	"end of listing",
	"start of parameter sequence",
	"end of parameter sequence",
	"frame","structure","item",
	"NEWLINE","INDENT","space","UNCOVER",
	"OVERLAY","ONLY","URL","column","column separator","end block","overlay range",
	"property", "comma", "part", "end of listing head", "description separater" };

/* ------------------------------------ *
 * local function declaration for       *
 * parsing and output generation        */
int parse_section();
int parse_frame();
int parse_title();
int parse_author();
int parse_date();
int parse_listing();
int parse_figure();
int parse_structure();
int parse_item();
int parse_enumerate();
int parse_properties();
int parse_frame_body();
int parse_url();
int parse_column();
int parse_column_sep();
int parse_end();
int parse_part();

/* ------------------------------------ */
/* global properties                    */

char *title, *author, *date; // properties for the whole document, could be used in future to generate the titlepage
int token;		// token id of the last fetched token
int structure_count = 0;
FILE *ofile;		   // the handle for the output file
char* data = NULL;        // data contains the result string from the scanner

/* Contains the name and type of the active item list:
 * itemize, enumeration, description
 */
char *item_type;

/* ------------------------------------ */
/* internal helper functions            */

/* Removes leading and trailing whitespaces.
 * The resulting string gets a new string object
 *
 * string = input
 * 
 * returns a string without trailing/leading whitespaces
 */
char* strtrim(const char *string) {
	while (isspace(*string))
		string++;
	char *result = strcpy((char*) malloc(strlen(string) + 1), string);
	char *last = result + strlen(result) - 1;
	while (isspace(*last))
		last--;
	last++;
	*last = 0;
	return result;
}

/* Displays the token name or token description for a
 * given token from the symbol table.
 *
 * token_id = numerical value representing the token
 *
 * returns a string
 */
const char* get_token_name(int token_id) {
	return token_names[token_id-101];
}

/* ------------------------------------ */
/* parser functions                     */

/* Main parser function
 */
int parse() {
	token = yylex();
	do {
		switch (token) {
		case TITLE:
			if (!parse_title()) return 0;
			break;
		case AUTHOR:
			if (!parse_author()) return 0;
			break;
		case DATE:
			if (!parse_date()) return 0;
			break;
		case SECTION:
			if (!parse_section()) return 0;
			break;
		case FRAME:
			if (!parse_frame()) return 0;
			break;
		case PART:
			if (!parse_part()) return 0;
			break;
		case EOF:
		case ZERO:
			return 1;
		case NEWLINE:
			token = yylex();
			break;
		case IMAGE:
			fprintf(stderr, "[%d] Top mode: figure not allowed in top mode.\n",yylineno);
			return 0;
		case LISTING:
			fprintf(stderr, "[%d] Top mode: listing not allowed in top mode.\n", yylineno);
			return 0;
		case STRUCT:
			fprintf(stderr, "[%d] Top mode: structure not allowed in top mode.\n", yylineno);
			return 0;
		case ITEM:
			fprintf(stderr, "[%d] Top mode: item not allowed in top mode.\n", yylineno);
			return 0;
		default:
			fprintf(stderr, "[%d] Top mode: %s \"%s\" not allowed in top mode.\n", yylineno,
					get_token_name(token), yytext);
			return 0;
		}
	} while (token != EOF);

	return 1;
}


/*
 * parse part rule
 */
int parse_part() {
	if ((token = yylex()) == VALUE) {
		fprintf(ofile,"\n\n%%%%%%\n\\part{%s}\n\n", data);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "[%d] Part value expected, but %s \"%s\" found\n", yylineno, get_token_name(token),yytext);
		return 0;
	}
}

/*
 * parse section rule
 */
int parse_section() {
	if ((token = yylex()) == VALUE) {
		fprintf(ofile,"\n%%%%\n\\section{%s}\n\n", data);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "[%d] Section value expected, but %s found\n", yylineno, get_token_name(token));
		return 0;
	}
}

/*
 * parse the content of a frame
 */
int parse_frame() {
	int subslide = 1;

	structure_count = 0;

	fprintf(ofile,"%% FRAME\n");
	fprintf(ofile,"\\begin{frame}[fragile]\n");

	if ((token = yylex()) == VALUE) {
		fprintf(ofile,"\\frametitle{%s}\n", data);
		token = yylex();
		while (token != EOF) {
			switch (token) {
			case UNCOVER:
				fprintf(ofile,"\\uncover<%d->{",subslide++);
				if (!parse_frame_body()) return 0;
				fprintf(ofile,"}\n");
				break;
			case OVERLAY:
				fprintf(ofile,"\\overlay<%d->{",subslide++);
				if (!parse_frame_body()) return 0;
				fprintf(ofile,"}\n");
				break;
			case ONLY:
				fprintf(ofile,"\\only<%d>{",subslide++);
				if (!parse_frame_body()) return 0;
				fprintf(ofile,"}\n");
				break;
			case STRUCT:
				if (!parse_structure()) return 0;
				break;
			case IMAGE:
				if (!parse_figure()) return 0;
				break;
			case LISTING:
				if (!parse_listing()) return 0;
				break;
			case ITEM:
				if (parse_item(1,item_type) == -1) return 0;
				break;
			case URL:
				if (!parse_url()) return 0;
				break;
			case COLUMN:
				if (parse_column() == -1) return 0;
				break;
			case COLUMN_SEP:
				if (parse_column_sep() == -1) return 0;
				break;
			case END:
				if (parse_end() == -1) return 0;
				break;
			case FRAME:
			case SECTION:
			case TITLE:
			case AUTHOR:
			case DATE:
			case ZERO:
			case PART:
				fprintf(ofile,"\\end{frame}\n\n");
				return 1;
			case NEWLINE: // ignore empty newline
				token = yylex();
				break;
			default:
				fprintf(stderr, "[%d] Frame mode: illegal token %s found.\n", yylineno, get_token_name(token));
				fprintf(stderr, "Expected commands: uncover (+), overlay (#), only (~), structure ('), image, listing, item or URL\n");
				fprintf(stderr, "\tor end frame mode with: New frame, section, title, author, date or an end of file\n");
				return 0;
			}
		}
		return 1;
	} else {
		fprintf(stderr, "[%d] Frame mode: expected value, but %s found.\n", yylineno, get_token_name(token));
		return 0;
	}
}

/* parser and renderer for nested frames
 */
int parse_frame_body() {
	token = yylex();
	switch (token) {
	case STRUCT:
		return parse_structure();
	case IMAGE:
		return parse_figure();
	case LISTING:
		return parse_listing();
	case ITEM:
		return parse_item(1,item_type);
	case URL:
		return parse_url();
	default:
		fprintf(stderr,"[%d] Overlay mode: Illegal token %s\n", yylineno, get_token_name(token));
		fprintf(stderr,"Expected commands: structure ('), image, listing, item and URL\n");
		return 0;
	}
}

/* parser and renderer for start of two columns
 */
int parse_column() {
	fprintf(ofile,"\\begin{columns}[t]\n\\begin{column}{0.5\\textwidth}\n");
	token = yylex();
	return 0;
}

/* parser and renderer for seperator of two columns
 */
int parse_column_sep() {
	fprintf(ofile,"\\end{column}\n\\begin{column}{0.5\\textwidth}\n");
	token = yylex();
	return 0;
}

/* parser and renderer for end of two columns
 */
int parse_end() {
	fprintf(ofile,"\\end{column}\n\\end{columns}\n");
	token = yylex();
	return 0;
}


int parse_url() {
	if ((token = yylex()) == VALUE) {
		fprintf(ofile,"\\url{%s}\n",data);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "[%d] Title value expected, but %s found.\n", yylineno, get_token_name(token));
		return 0;
	}
}

/*
 *
 */
int parse_title() {
	if ((token = yylex()) == VALUE) {
		title = data;
		fprintf(stderr, "Title: %s\n", title);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "[%d] Title value expected, but %s found.\n", yylineno, get_token_name(token));
		return 0;
	}
}

/*
 *
 */
int parse_author() {
	if ((token = yylex()) == VALUE) {
		author = data;
		fprintf(stderr, "Author: %s\n", author);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "[%d] Author value expected, but %s found.\n", yylineno, get_token_name(token));
		return 0;
	}
}

/*
 *
 */
int parse_date() {
	if ((token = yylex()) == VALUE) {
		date = data;
		fprintf(stderr, "Date: %s\n", date);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "[%d] Date value expected, but %s found.\n", yylineno, get_token_name(token));
		return 0;
	}
}

/*
 * process listings
 */
int parse_listing_text() {
	while ((token = yylex()) == LINE) {
		fprintf(ofile,"%s", data);
	}
	if (token != LSTEND) {
		fprintf(stderr,
				"[%d] Listing mode: End of listing or listing expected, but %s \"%s\" found\n",
				yylineno, get_token_name(token), yytext);
		return 0;
	} else {
		fprintf(ofile,"\\end{lstlisting}\n");
		token = yylex();
		return 1;
	}
}


int parse_listing() {
	fprintf(ofile,"\\begin{lstlisting}");
	if ((token = yylex()) == ID) {
		fprintf(ofile,"[escapechar=@,language=%s", data);
		if ((token = yylex()) == SEP) {
			if ((token = yylex()) == LABEL) {
				fprintf(ofile,",caption=%s]\n", data);
				return parse_listing_text();
			} else {
				fprintf(stderr, "[%d] Listing mode: missing label for listing, found %s \"%s\" instead\n", yylineno, get_token_name(token), yytext);
				return 0;
			}
		} else if (token == HEAD_END) {
			fprintf(ofile,"]\n");
			return parse_listing_text();
		} else {
			fprintf(stderr, "[%d] Listing mode: Separator '-' expected\n", yylineno);
			return 0;
		}
	} else {
		fprintf(stderr, "[%d] Listing mode: Language name expected, but %s \"%s\"\n", yylineno, get_token_name(token), yytext);
		return 0;
	}
}

/*
 * process figures
 */
int parse_figure_inner(int index) {
	char *overlay_code = "";

	token = yylex();
	if (token == OVERLAY_CODE) {
		overlay_code = data;
		token = yylex();
	}
	if (token == PSEQ_START) {
		fprintf(ofile,"\\includegraphics%s[",overlay_code);
		if (parse_properties() == 0)  {
			if ((token = yylex()) == VALUE) {
				fprintf(ofile,"]{%s}\n", data);
				token = yylex();
				if (token == IMAGE) // is an image group
					parse_figure_inner(index+1);
				return 0;
			} else {
				fprintf(stderr, "[%d] Figure mode: Missing image, found %s instead\n", yylineno, get_token_name(token));
				return -1;
			}
		} else
			return -1;
	} else {
		fprintf(stderr, "[%d] Figure mode: Missing [, found %s instead\n", yylineno, get_token_name(token));
		return -1;
	}
}

int parse_figure() {
	fprintf(ofile,"\\begin{figure}\n");
	if (parse_figure_inner(0) == 0) {
		fprintf(ofile,"\\end{figure}\n");
		return 1;
	} else
		return 0;
}

/*
 * parse properties
 */
int parse_properties() {
	while ((token = yylex()) != PSEQ_END) {
		switch (token) {
		case PROPERTY:
			fprintf(ofile,"%s",data);
			break;
		case PROP_SEP:
			fprintf(ofile,",");
			break;
		case PSEQ_END:
			break;
		default:
			fprintf(stderr, "[%d] Figure mode, property sequence: Missing ]. Found %s\n", yylineno,get_token_name(token));
			return -1;
		}
	}
	return 0;
}

/*
 *
 */
int parse_structure() {
	if ((token = yylex()) == VALUE) {
		if (structure_count > 0)
			fprintf(ofile,"\\vskip1em\n");
		structure_count++;
		fprintf(ofile,"\\structure{%s}\n", data);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "[%d] Figure mode: Missing ]\n", yylineno);
		return 0;
	}
}


int parse_item_value(const char *mode) {
	if (strcmp("description",mode) == 0) {
		if ((token = yylex()) == LABEL) {
			fprintf(ofile,"\\item[%s] ",data);
			if ((token = yylex()) != DESC_SEP) {
				fprintf(stderr, "<%s:%d> [%d] Description item mode: Label-value-separator = expected, but %s \"%s\" found instead.\n",__FILE__,__LINE__, yylineno,get_token_name(token),yytext);
				return 0;
			} else if ((token = yylex()) == VALUE) {
				fprintf(ofile,"%s\n",data);
				token = yylex();
				return 1;
			} else {
				fprintf(stderr, "<%s:%d> [%d] Item mode: Missing value for item, got %s \"%s\" instead.\n",__FILE__,__LINE__, yylineno,get_token_name(token),yytext);
				return 0;
			}
		} else {
			fprintf(stderr, "<%s:%d> [%d] Description item mode: Missing label for item, got %s \"%s\" instead.\n",__FILE__,__LINE__, yylineno,get_token_name(token),yytext);
			return 0;
		}
	} else if ((token = yylex()) == VALUE) {
		fprintf(ofile,"\\item %s\n",data);
		token = yylex();
		return 1;
	} else {
		fprintf(stderr, "<%s:%d> [%d] Item mode: Missing value for item, got %s \"%s\" instead.\n",__FILE__,__LINE__, yylineno,get_token_name(token),yytext);
		return 0;
	}
}

/*
 *
 */
int parse_item(int indent, char *mode) {
	if (strcmp("description",mode) == 0) {
		if ((token = yylex()) == LABEL) {
			fprintf(ofile,"\\begin{%s}\n",mode);
			fprintf(ofile,"\\item[%s] ",data);
			if ((token = yylex()) != DESC_SEP) {
				fprintf(stderr, "<%s:%d> [%d] Description item mode: Label-value-separator = expected, but %s \"%s\" found instead.\n",__FILE__,__LINE__, yylineno,get_token_name(token),yytext);
				return -1;
			} else if ((token = yylex()) == VALUE) {
				fprintf(ofile,"%s\n",data);
			}
		} else {
			fprintf(stderr, "<%s:%d> [%d] Description item mode: Missing label for item, got %s \"%s\" instead.\n",__FILE__,__LINE__, yylineno,get_token_name(token),yytext);
			return -1;
		}
	} else if ((token = yylex()) == VALUE) { // first value
		fprintf(ofile,"\\begin{%s}\n",mode);
		fprintf(ofile,"\\item %s\n",data);
	} else {
		fprintf(stderr, "<%s:%d> [%d] Item mode: Missing value for item, got %s \"%s\" instead\n",__FILE__,__LINE__, yylineno,get_token_name(token),yytext);
		return -1;
	}

	token = yylex();
	do {
		if (token == INDENT) {
			int indent_length = strlen(data)+1;
			if (indent==indent_length) { // same indent
				if ((token = yylex()) == ITEM) {
					if (!parse_item_value(mode)) return -1;
				} else {
					fprintf(stderr,"<%s:%d> [%d] Item expected, but %s \"%s\" found.\n",__FILE__,__LINE__,yylineno,get_token_name(token),yytext);
					return -1;
				}
			} else if (indent>indent_length) { // less indent go back
				fprintf(ofile,"\\end{%s}\n",mode);
				return indent_length;
			} else if (indent<indent_length) { // new indent go sub
				if ((token = yylex()) == ITEM) {
					int result = parse_item(indent+1,item_type);
					if (result > 0) { // not an error
						if (result < indent) {
							fprintf(ofile,"\\end{%s}\n",mode);
							return result;
						} else	if (result == indent) {
							if (!parse_item_value(mode)) return -1;
						}
					} else if (result == -1) {
						return -1;
					} else {
						if (result < indent) 
							fprintf(ofile,"\\end{%s}\n",mode);
						return 0;
					}
				} else {
					fprintf(stderr,"<%s:%d> [%d] Item expected, but %s \"%s\" found.\n",__FILE__,__LINE__,yylineno,get_token_name(token),yytext);
					return -1;
				}
			}
		} else {
			switch (token) {
			case ITEM: // go back all indent, base item list	
				if (indent>1) {	
					fprintf(ofile,"\\end{%s}\n",mode);
					return 1;
				} else if (!parse_item_value(mode)) 
					return -1;
				break;
			case SECTION:
			case FRAME:
			case IMAGE:
			case STRUCT:
			case NEWLINE:
			case EOF:
			case COLUMN:
			case COLUMN_SEP:
			case END:
			case ZERO: // leave frame => close all item list
				fprintf(ofile,"\\end{%s}\n",mode);
				return 0;
			default:
				fprintf(stderr, "<%s:%d> [%d] Top mode: %s \"%s\" not allowed in top mode.\n",__FILE__,__LINE__,
										yylineno, get_token_name(token), yytext);
				fprintf(stderr, "<%s:%d> Expected: item, indent, section, frame, image, structure ('), newline, or end of file\n",__FILE__,__LINE__);
				return -1;
			}
		}
	} while (token != EOF);
	fprintf(stderr, "<%s:%d> [%d] Item mode: Unhandled EOF\n",__FILE__,__LINE__, yylineno);
	return -1;
}


